File: backend-fast-api/controllers/auth_controller.py
from fastapi import HTTPException
from passlib.context import CryptContext
from jose import jwt
from dotenv import load_dotenv
from config.db import get_db
import os
from bson import ObjectId  

load_dotenv()
JWT_SECRET = os.getenv("JWT_SECRET", "your-secret-key-here")  # Default to a fallback if not set
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

def generate_token(user_id: str):
    payload = {"id": user_id}
    return jwt.encode(payload, JWT_SECRET, algorithm="HS256")

async def register_user(username: str, password: str):
    db = await get_db()  
    users_collection = db["users"]
    if await users_collection.find_one({"username": username}):  
        raise HTTPException(status_code=400, detail="User already exists")
    
    hashed_password = pwd_context.hash(password)
    user = {"username": username, "password": hashed_password}
    result = await users_collection.insert_one(user) 
    user_id = str(result.inserted_id)
    return {"token": generate_token(user_id), "user": {"id": user_id, "username": username}}

async def login_user(username: str, password: str):
    db = await get_db() 
    users_collection = db["users"]
    user = await users_collection.find_one({"username": username})  
    if not user or not pwd_context.verify(password, user["password"]):
        raise HTTPException(status_code=400, detail="Invalid credentials")
    user_id = str(user["_id"])
    return {"token": generate_token(user_id), "user": {"id": user_id, "username": username}}

File: backend/controllers/authController.js
const User = require("../models/User");
const jwt = require("jsonwebtoken");

const generateToken = (user) => {
  return jwt.sign({ id: user._id }, process.env.JWT_SECRET, {
    expiresIn: "1h",
  });
};

const registerUser = async (req, res) => {
  try {
    const { username, password } = req.body;
    const existingUser = await User.findOne({ username });
    if (existingUser) {
      return res.status(400).json({ message: "User already exists" });
    }

    const user = await User.create({ username, password });
    res.status(201).json({ token: generateToken(user), user });
  } catch (error) {
    res.status(500).json({ message: "Error registering user", error });
  }
};


const loginUser = async (req, res) => {
  try {
    const { username, password } = req.body;
    const user = await User.findOne({ username });

    if (!user || !(await user.comparePassword(password))) {
      return res.status(400).json({ message: "Invalid credentials" });
    }

    res.json({ token: generateToken(user), user });
  } catch (error) {
    res.status(500).json({ message: "Error logging in", error });
  }
};

module.exports = { registerUser, loginUser };
